### PR TITLE
Add gas to base operation (tx, roll buy & sell)

### PR DESCRIPTION
**Pull request by bilboquet**
Fixes https://github.com/massalabs/massa/issues/4467

* [x] document all added functions
* [x] try in sandbox /simulation/labnet
  * [x] if part of node-launch, checked using the `resync_check` flag
* [x] unit tests on the added/changed features
  * [x] make tests compile
  * [x] make tests pass 
* [x] add logs allowing easy debugging in case the changes caused problems
* [x] if the API has changed, update the API specification

### CODE REVIEW

File: massa-models/src/config/constants.rs [hunk 1]
```diff
@@ -240,6 +240,8 @@ pub const POOL_CONTROLLER_DENUNCIATIONS_CHANNEL_SIZE: usize = 1024;
 pub const MAX_GAS_PER_BLOCK: u64 = u32::MAX as u64;
 /// Maximum of GAS allowed for asynchronous messages execution on one slot
 pub const MAX_ASYNC_GAS: u64 = 1_000_000_000;
+/// Gas used by a base operation (transaction, roll buy, roll sell)
+pub const BASE_OPERATION_GAS_COST: u64 = MAX_GAS_PER_BLOCK / (MAX_OPERATIONS_PER_BLOCK as u64);
```
Comment (damip) [inline massa-models/src/config/constants.rs:244]: use a constant here directly

File: massa-models/src/operation.rs [hunk 2]
```diff
@@ -972,13 +970,13 @@ impl SecureShareOperation {
     }
 
     /// Get the max amount of gas used by the operation (`max_gas`)
-    pub fn get_gas_usage(&self) -> u64 {
+    pub fn get_gas_usage(&self, gas_usage: u64) -> u64 {
```
Comment (damip) [inline massa-models/src/operation.rs:973]: * rename "gas_usage" to "base_operation_gas_cost" to be uniform with the settings
* also apply `base_op_gas_usage` to executesc and callsc: `max_gas.saturating_add(base_op_gas_usage)`

File: massa-models/src/operation.rs [hunk 3]
```diff
@@ -972,13 +970,13 @@ impl SecureShareOperation {
     }
 
     /// Get the max amount of gas used by the operation (`max_gas`)
```
Comment (damip) [inline massa-models/src/operation.rs:972]: update the docs

File: massa-models/src/operation.rs [hunk 4]
```diff
@@ -972,13 +970,13 @@ impl SecureShareOperation {
...
-    pub fn get_gas_usage(&self) -> u64 {
+    pub fn get_gas_usage(&self, gas_usage: u64) -> u64 {
         match &self.content.op {
             OperationType::ExecuteSC { max_gas, .. } => *max_gas,
             OperationType::CallSC { max_gas, .. } => *max_gas,
```
Comment (damip) [inline massa-models/src/operation.rs:976]: also apply `base_op_gas_usage` to executesc and callsc: `max_gas.saturating_add(base_op_gas_usage)`

Comment (damip) [reply massa-models/src/operation.rs:976]: even better:

```
match ... {
   executesc => max_gas,
   callsc => max_gas
   rollbuy => 0
   rollsell => 0
   transaction => 0
}.saturating_add(base_op_gas_usage)
```

File: massa-pool-exports/src/config.rs [hunk 5]
```diff
@@ -13,6 +13,8 @@ pub struct PoolConfig {
     pub max_block_size: u32,
     /// maximal gas per block
     pub max_block_gas: u64,
+    /// Gas used by a base operation (transaction, roll buy, roll sell)
```
Comment (damip) [inline massa-pool-exports/src/config.rs:16]: it applies to all types of ops (not only rollbuy/rollsell/tx)

File: massa-models/src/config/constants.rs (hunk 1 above)
Comment (bilboquet) [reply massa-models/src/config/constants.rs]: done (but I found it better expressed relatively to other constants)

File: massa-models/src/operation.rs (hunk 3 above)
Comment (bilboquet) [reply massa-models/src/operation.rs]: done

File: massa-models/src/operation.rs (hunk 4 above)
Comment (bilboquet) [reply massa-models/src/operation.rs:979]: done.
@Eitu33: any conflict with your work here?

File: massa-pool-exports/src/config.rs (hunk 5 above)
Comment (bilboquet) [reply massa-pool-exports/src/config.rs]: changed the comment

File: massa-models/src/operation.rs (hunk 4 above)
Comment (Eitu33) [reply massa-models/src/operation.rs:979]: Not really, max_gas will be the same for OP and ABI and base_op_gas_usage will even out abi_cost. The total cost won't be the exact same but it's not an important difference at all.

Comment (bilboquet) [reply massa-models/src/operation.rs:979]: Great, thanks

File: massa-models/src/config/constants.rs (hunk 1 above)
Comment (damip) [reply massa-models/src/config/constants.rs]: I'm scared of "unexpected correlations. Eg. if we change the max_block_gas we don't expect the base cost of ops to be affected

**Review by damip — APPROVED** [review]

**Review by Eitu33 — APPROVED** [review]

**Comment by bilboquet** [comment]
bors merge

**Review by Leo-Besancon — APPROVED** [review]

**Pull request merged**